src-tauri/src/recorder/encoder/mac/enc-swift/src/lib.swift

```diff
@@ -0,0 +1,248 @@
+import SwiftRs
+import Foundation
+import AVFoundation
+import VideoToolbox
+import CoreVideo
+import CoreImage
+
+// TODO: handle errors better
+class Encoder: NSObject {
+    var width: Int
+    var height: Int
+    var assetWriter: AVAssetWriter
+    var assetWriterInput: AVAssetWriterInput
+    var pixelBufferAdaptor: AVAssetWriterInputPixelBufferAdaptor
+
+    init(_ width: Int, _ height: Int, _ outFile: URL) {
+        self.width = width;
+        self.height = height;
+
+        // Setup AVAssetWriter
+        // Create AVAssetWriter for a mp4 file
+        self.assetWriter = try! AVAssetWriter(url: outFile, fileType: .mp4)
+        
+        // Prepare the AVAssetWriterInputPixelBufferAdaptor
+        let outputSettings: [String: Any] = [
+            AVVideoCodecKey: AVVideoCodecType.h264,
+            AVVideoWidthKey: width,
+            AVVideoHeightKey: height
+        ]
+
+        self.assetWriterInput = AVAssetWriterInput(mediaType: .video, outputSettings: outputSettings)
+        self.assetWriterInput.expectsMediaDataInRealTime = true
+
+        let sourcePixelBufferAttributes: [String: Any] = [
+            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_420YpCbCr8BiPlanarFullRange
+        ]
+
+        self.pixelBufferAdaptor = AVAssetWriterInputPixelBufferAdaptor(
+            assetWriterInput: self.assetWriterInput,
+            sourcePixelBufferAttributes: sourcePixelBufferAttributes
+        )
+        
+        if self.assetWriter.canAdd(self.assetWriterInput) {
+            self.assetWriter.add(self.assetWriterInput)
+        }
+
+        self.assetWriter.startWriting()
+        self.assetWriter.startSession(atSourceTime: CMTime.zero)
+    }  
+}
+
+
+@_cdecl("encoder_init")
+func encoderInit(_ width: Int, _ height: Int, _ outFile: SRString) -> Encoder {
+    return Encoder(
+        width,
+        height,
+        URL(fileURLWithPath: outFile.toString())
+    )
+}
+
+func createCvPixelBufferFromYuvFrameData(
```

Comment:
this is a util func for  encoderIngestYuvFrame
can be moved to separate file. Similarly util func for bgra as well.

Reply:
solved here: https://github.com/helmerapp/henx/commit/b6a774deca7a3d95f31d6192517805493bc3e188